src/main/java/net/minecraftforge/common/CoordFlyweightSet.java

```diff
@@ -0,0 +1,192 @@
+ * 
+ * Such Mods may degrade the performance gained by this optimization.
+ */
+public class CoordFlyweightSet implements Set<ChunkCoordIntPair> {
+    private final TLongHashSet packedSet = new TLongHashSet(32);
+
+    @Override
+    public int size()
+    {
+        return packedSet.size();
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return packedSet.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o)
+    {
+        if(o instanceof ChunkCoordIntPair) {
+            ChunkCoordIntPair intPair = (ChunkCoordIntPair)o;
+            long packed = ForgeChunkManager.packedChunkCoordIntPair(intPair.chunkXPos,intPair.chunkZPos);
+            return containsPacked(packed);
+        }
+        
+        return false;
+    }
+    
+    public boolean containsPacked(long packedCoord) {
+        return packedSet.contains(packedCoord);
+    }
+
+    @Override
+    public java.util.Iterator<ChunkCoordIntPair> iterator()
+    {
+        return new CoordFlyweightSet.Iterator();
+    }
+    
+    public TLongIterator packedIterator() {
+        return packedSet.iterator();
+    }
+
+    @Override
+    public Object[] toArray()
+    {
+        return toArray(new Object[size()]);
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a)
+    {
+        int i = 0;
+        for(ChunkCoordIntPair fat : this) {
+            a[i++] = (T)fat;
+        }
+        return a;
+    }
+
+    @Override
+    public boolean add(ChunkCoordIntPair e)
+    {
+        return add(e.chunkXPos, e.chunkZPos);
+    }
+    
+    public boolean add(int chunkXPos, int chunkZPos) {
+        return packedSet.add(
+                ForgeChunkManager.packedChunkCoordIntPair(chunkXPos, chunkZPos));
+    }
+    
+    public boolean addPacked(long intPair) {
+        return packedSet.add(intPair);
+    }
+
+    @Override
+    public boolean remove(Object o)
+    {
+        if(o instanceof ChunkCoordIntPair) {
+            ChunkCoordIntPair intPair = (ChunkCoordIntPair)o;
+            return packedSet.remove(
+                    ForgeChunkManager.packedChunkCoordIntPair(intPair.chunkXPos, intPair.chunkZPos));
+        }
+        
+        return false;
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> c)
+    {
+        for(Object o : c) {
+            if(!contains(o)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends ChunkCoordIntPair> c)
+    {
+        boolean changed = false;
+        for(ChunkCoordIntPair intPair : c) {
+            changed |= add(intPair);
+        }
+        return changed;
+    }
+
+    @Override
+    public boolean retainAll(Collection<?> c)
+    {
+        if(c instanceof CoordFlyweightSet) {
+            return packedSet.retainAll(((CoordFlyweightSet)c).packedSet);
+        }
+        
+        long[] retain = null;
```

Comment:
Going to rewrite this. It inappropriately retains 0L if Collection contains null entries.